src/file_watching.jl

```diff
@@ -1,56 +1,153 @@
+        end
     end
-    # fire user-specified callback
-    fw.callback(filepath, fileevent)
 end
 
 """
-    add_to_filewatcher!(filewatcher, filepath)
+    _waitfor_task_shutdown(w::SimpleWatcher)
+
+[INTERNAL] Helper function ensuring that the `_file_watcher` task has ended
+before continuing.
+"""
+_waitfor_task_shutdown(w::SimpleWatcher) = while !istaskdone(w.task) sleep(0.05) end
+
+"""
+    set_callback(w::SimpleWatcher, fcn::Function)
+
+API function to set or change the callback function being executed upon a
+file change. Can be "hot-swapped", i.e. while the file watcher is running.
+The callback function receives a string with the file path and is not
+expected to return anything.
+"""
+function set_callback(w::SimpleWatcher, fcn::Function)
+    prev_running = stop(w) # returns true if was running
+    w.callback = fcn
+    prev_running && start(w) # start again if it was running before
+end
+
+"""
+    isrunning(w::SimpleWatcher)
+
+API function to check whether the file watcher is running. Should not be
+required though, since the `start` and `stop` API functions work in any
+case.
+"""
+isrunning(w::SimpleWatcher) = (w.task != nothing) && !istaskdone(w.task)
+
+"""
+    start(w::SimpleWatcher)
+
+API function to start the file watcher.
+"""
+start(w::SimpleWatcher) = !isrunning(w) && (w.task = @async _file_watcher(w))
+
+"""
+    stop(w::SimpleWatcher)
 
-Adds `filepath` to the FileWatcher `filewatcher` provided it is not already being watched.
+API function to stop the file watcher. The list of files being watched is
+preserved. Also, it still accepts new files to be watched by `watch_file`.
+Once restarted with `start`, it continues watching the files in the list
+(and will initially trigger the callback for all files that have changed
+in the meantime). Returns a `Bool` indicating whether the watcher was
+running before `stop` was called.
 """
-function add_to_filewatcher!(fw::FileWatcher, filepath::AbstractString)
-    if filepath ∉ keys(fw.watched_files)
-        fw.watched_files[filepath] = @async _file_watcher(filepath, fw)
+function stop(w::SimpleWatcher)
+    was_running = isrunning(w)
+    if was_running
+        schedule(w.task, InterruptException(), error=true)
+        _waitfor_task_shutdown(w) # required to have consistent behaviour
     end
-    return nothing
+    return was_running
 end
 
+
 """
-    stop_tasks!(filewatcher)
+    is_file_watched(w::SimpleWatcher, path::String)
 
-Kill all file-watching tasks of a `FileWatcher`."
+API function to check whether a file is already being watched.
 """
-function stop_tasks!(fw::FileWatcher)
-    for (fp, tsk) ∈ fw.watched_files
-        if tsk.state == :runnable
-            schedule(tsk, [], error=true) # raise error in the task --> kills it
-        end
+# is_file_watched(w::SimpleWatcher, path::String) = path ∈ [wf.path for wf ∈ w.filelist]
+function is_file_watched(w::SimpleWatcher, path::String)
+    return path ∈ [wf.path for wf ∈ w.filelist]
```

Comment:
you can probably use a generator here instead of builing a vector of paths.

Reply:
Generator?
Could use `mapreduce`:
```julia
mapreduce(wf -> wf.path == path, |, w.filelist)
```

Reply:
Hehe, did some benchmarking just for fun:
```julia
using LiveServer: WatchedFile
using BenchmarkTools: @btime

fl = [WatchedFile("a"), WatchedFile("b"), WatchedFile("c")]

println("array:")
@btime "b" ∈ [wf.path for wf ∈ fl]
@btime "d" ∈ [wf.path for wf ∈ fl]

println("mapreduce:")
@btime mapreduce(wf -> wf.path == "b", |, fl)
@btime mapreduce(wf -> wf.path == "d", |, fl)
println("reduce:")
@btime reduce((old, wf) -> old || (wf.path == "b"), fl, init=false)
@btime reduce((old, wf) -> old || (wf.path == "d"), fl, init=false)

function is_watched(fl, path)
    for f ∈ fl f.path == path && return true end
    return false
end

println("loop:")
@btime is_watched(fl, "b")
@btime is_watched(fl, "d")
```
The results:
```
array:
  101.822 ns (2 allocations: 128 bytes)
  112.910 ns (2 allocations: 128 bytes)
mapreduce:
  186.651 ns (0 allocations: 0 bytes)
  181.470 ns (0 allocations: 0 bytes)
reduce:
  136.187 ns (2 allocations: 48 bytes)
  139.678 ns (2 allocations: 48 bytes)
loop:
  18.879 ns (0 allocations: 0 bytes)
  20.879 ns (0 allocations: 0 bytes)
```
So the clear winner is... the good 'ol loop! It's also clear what it does since no fancy map'n'reduce stuff is used. So use that one?

Reply:
`path ∈ [wf.path for wf ∈ w.filelist]` can be written (and this doesn't allocate)

```julia
any(wf -> path == wf.path, w.filelist)
```

Reply:
Ok, that's clean. Not that we have to care about nanoseconds here, but adding it to the benchmarks:
```
println("any:")
@btime any(wf -> wf.path == "b", fl)
@btime any(wf -> wf.path == "d", fl)
```
yields:
```
any:
  127.790 ns (0 allocations: 0 bytes)
  139.329 ns (0 allocations: 0 bytes)
```
Wondering why the loop is so much faster (than everything else). Will go with `any` and keep the fast version in a comment (if it ever should make a difference).

Reply:
You should watch out because you don't use interpolation for ambient variables in your benchmark so potentially the timings are wrong. A simple fix would be to wrap all of them in functions. But it's irrelevant anyway I think here readability wins.

Reply:
Couldn't leave it alone; here's everything wrapped in a function and with 10s sleep between every benchmark to "cool down" the CPU again:
```
array:
  50.966 ns (2 allocations: 128 bytes)
  54.673 ns (2 allocations: 128 bytes)
mapreduce:
  25.753 ns (1 allocation: 16 bytes)
  24.408 ns (1 allocation: 16 bytes)
reduce:
  23.596 ns (1 allocation: 16 bytes)
  24.177 ns (1 allocation: 16 bytes)
loop:
  16.683 ns (0 allocations: 0 bytes)
  20.326 ns (0 allocations: 0 bytes)
any:
  22.363 ns (1 allocation: 16 bytes)
  23.775 ns (1 allocation: 16 bytes)
```
So, except for explicitly building the array, everything is roughly the same.

Reply:
yeah I think the no allocations ones should basically kind of amount to the same lowered code which would amount to a lazy loop  (i.e. the same loop that you did but without forming the array).